### PR TITLE
[ROCm] Make DSA indexer top-k exact with cooperative selection

### PR DESCRIPTION
**Pull request by jiejingzhangamd**
## Summary

- add a ROCm-specific cooperative top-k implementation for the three DSA indexer entry points
- use an fp32 coarse histogram, exact radix tie refinement, and an overflow rescan
- split long rows across blocks when a low batch size would otherwise leave the GPU underutilized
- preserve the raw, paged, and ragged interfaces and guard launches to the score device
- keep the CUDA implementation unchanged; ROCm tracks the native `topk.hip` source and does not hipify `topk.cu`

## Motivation

The previous ROCm path used `topk.hip` generated by hipifying `topk.cu`. Its fp16 coarse key can collapse a large part of a real indexer row into one bucket, and overflowing the candidate buffer can silently drop valid winners. On captured GLM-5.2 logits, selected-value recall fell as low as 0.53.

The new native HIP kernel keeps the selection exact by refining the full fp32 key and rescanning the row when the tie buffer overflows. It reaches 1.0 selected-value recall against `torch.topk` on the same inputs.

## Performance vs. the original `topk.hip`

Baseline `old generated` is the original ROCm `topk.hip` produced from `topk.cu`, rebuilt from xiaobochen-amd/sglang `main` at [`92562f066`](https://github.com/xiaobochen-amd/sglang/commit/92562f0660b1c05d2de5b823e8b471be8def44cd). `new native` is the tracked `topk.hip` in this PR. Measurements are on MI355X (gfx950), graph replay, width 131,072, 100,500 live positions, top-k 2,048, page size 1. Times are microseconds per call; lower is better.

```
distribution  batch   old generated   new native   speedup
tiny              1            95.2         31.6      3.02x
tiny              2            95.5         30.0      3.18x
tiny              4            95.6         31.2      3.06x
tiny              8            96.0         31.4      3.06x
tiny             14            95.8         32.7      2.93x
tiny             32            96.3         36.9      2.61x
tiny             64            98.4         48.1      2.04x
tiny            128           119.9         58.5      2.05x
tiny            256           121.9         62.7      1.94x

diffuse           1            54.3         31.6      1.72x
diffuse           2            55.1         30.9      1.78x
diffuse           4            55.5         31.5      1.76x
diffuse           8            56.7         31.7      1.79x
diffuse          14            56.4         33.1      1.70x
diffuse          32            56.3         36.6      1.54x
diffuse          64            77.3         49.4      1.56x
diffuse         128            96.7         59.1      1.64x
diffuse         256           103.4         63.2      1.64x

clustered         1           160.9        102.2      1.57x
clustered         2           161.6        101.0      1.60x
clustered         4           161.8        102.1      1.59x
clustered         8           162.1        103.4      1.57x
clustered        14           162.0        104.9      1.54x
clustered        32           162.7        111.4      1.46x
clustered        64           162.6        122.9      1.32x
clustered       128           185.9        169.0      1.10x
clustered       256           186.9        174.2      1.07x
```

On captured GLM-5.2 decode shapes, row splitting gives a further 2.1-2.8x speedup over the original generated kernel.

## Comparison with `topk_v2`

This is a same-run comparison against the current JIT [`topk_v2.cuh`](https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh) on MI355X (gfx950). Both kernels use graph replay, width 131,072, 100,500 live positions, top-k 2,048, and page size 1. Times are microseconds per call; lower is better.

### GLM-5 ROCm shape and score distribution

The synthetic sweep uses a GLM-5 long-context decode-like shape: float32 scores `[B, 131072]`, live length 100,500, top-k 2,048, and a page-size-1 transform, with `B` swept from 1 to 256. Separate logits captured from an actual GLM-5.2-MXFP4 ROCm decode had 20 score rows per call and live context up to 134,849 tokens.

The real captured scores were approximately 0.04-180, not tiny `1e-16` values. They were nevertheless highly concentrated under the fp16 coarse key: an 8-bit fp16 key populated only 4-126 bins, and the largest bin held 6%-88% of a row. Whether the top-2,048 boundary landed inside that dominant bin varied by call.

The sweep distributions should therefore be read as follows:

- `diffuse`: broad random control; less concentrated than captured GLM-5 logits.
- `tiny`: scale-invariance stress (`1e-16 * randn`); not the literal GLM-5 value range.
- `clustered`: single-bin-style overflow stress; more concentrated than typical captures.

Real GLM-5 lies between `diffuse` and the synthetic `clustered` case in coarse-bin concentration. The actual captures are the reason candidate-buffer overflow must remain exact, while `tiny` separately verifies that fp16 conversion cannot erase score ordering.

```
distribution  batch   native topk.hip   topk_v2
tiny              1              31.5       73.1
tiny              2              30.1       73.2
tiny              4              30.8       73.1
tiny              8              30.9       73.2
tiny             14              32.5       73.2
tiny             32              37.1       73.3
tiny             64              48.8       73.3
tiny            128              58.7       74.6
tiny            256              63.2       76.1

diffuse           1              31.9       34.2
diffuse           2              31.0       34.2
diffuse           4              31.3       34.5
diffuse           8              31.6       34.5
diffuse          14              33.1       34.7
diffuse          32              36.6       35.1
diffuse          64              49.5       38.1
diffuse         128              59.2       39.6
diffuse         256              62.9       46.5

clustered         1             102.2      109.2
clustered         2             101.0      109.4
clustered         4             102.2      109.5
clustered         8             103.1      109.6
clustered        14             105.1      109.6
clustered        32             111.1      109.6
clustered        64             122.9      109.9
clustered       128             168.9      110.6
clustered       256             174.5      111.0
```

For tiny scores, the native HIP kernel is 1.20-2.43x faster. On diffuse scores it is 1.05-1.10x faster through batch 14, while `topk_v2` becomes 1.04-1.49x faster from batch 32 onward. For shorter diffuse rows (4K-65K live positions), `topk_v2` is 1.24-2.96x faster because its register and one-block streaming paths avoid the native kernel's row-split launch overhead.

Correctness is not equivalent. `topk_v2` uses an fp16 coarse histogram and caps the threshold-bin candidate buffer at 2,048 without an overflow rescan. Selected-value multiset recall against `torch.topk` at batch 1 is:

```
distribution   native topk.hip   topk_v2
tiny                    1.0000     0.0420
diffuse                 1.0000     1.0000
clustered               1.0000     0.0220
```

Therefore the apparent `topk_v2` advantage on high-batch clustered inputs is not a valid like-for-like speedup: it does not return the exact top-k on that distribution. The native HIP kernel rescans the row when the candidate buffer overflows and remains exact.

## Test plan

- [x] ROCm `sgl-kernel` build for gfx950 using the tracked native `topk.hip`
- [x] `pytest python/sglang/kernels/aot/tests/test_topk.py -q` - 161 passed
- [x] pre-commit checks on all changed files
- [x] mixed-device launch and rejection coverage
- [x] both one-block and row-split paths

---
### CI States

Latest PR Test (Base): :x: [Run #33953440606](https://github.com/sgl-project/sglang/actions/runs/33953440606)
Latest PR Test (Extra): :x: [Run #33953440438](https://github.com/sgl-project/sglang/actions/runs/33953440438)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33953440514](https://github.com/sgl-project/sglang/actions/runs/33953440514)

### CODE REVIEW

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
@1am9trash @RolaoDenthu

**Comment by HaiShaw** [comment]
HIP specific

**Comment by jiejingzhangamd** [comment]
@HaiShaw more compare with v2:
Comparison with topk_v2 on captured GLM-5.2 logits
Tested on MI355X with eight graph-replayed captures, each shaped [20, 134868], with live lengths 134849–134868 and top-k 2048.

Recall: native topk.hip was exact on all 160 rows. topk_v2 was exact on 159/160 rows; one row had 0.957 recall because its threshold bin contained 3296 candidates, exceeding the 2048-entry buffer.
Performance: on the seven exact captures, topk_v2 averaged 43.7 μs versus 44.7 μs for native topk.hip—about 2.2% faster. Across all captures, native averaged 45.3 μs versus 46.2 μs for topk_v2.
On the overflow capture, native topk.hip remained exact and ran in 49.4 μs, while topk_v2 ran in 64.0 μs—native was 1.30× faster.

**Comment by 1am9trash** [comment]
This PR reimplements `fast_topk` on the HIP platform. They are used by the indexer of `is_deepseek_dsa()` models (DeepSeek-V4 is unaffected).

image

The calling code path is `dsa_topk_backend.py`, which is what all DSA models except V4 go through. When `SGLANG_OPT_USE_TOPK_V2=1` it dispatches to the `topk_v2` kernel; otherwise it uses the one optimized in this PR.

Hi @jiejingzhangamd, 
I noticed that #36851 turned top-k v2 on by default for GLM on ROCm. Does that mean we never reach this kernel?
